tests/test_username_api.py

```diff
@@ -13,23 +12,31 @@
 data = yaml.load(open(os.path.join('tests', 'test_data.yml')))
 websites = yaml.load(open('websites.yml'))
 
-invalid_username = '$very%long{invalid}user(name)'
+AvailabilityTestCase = collections.namedtuple(
+    'AvailabilityTestCase',
+    'website username')
 
 
-def generate_random_username(website, length):
-    first_char_letter = False
-    if website == 'openhub':
-        first_char_letter = True
-    return ''.join(generate_random_character(length,
-                                             first_char_letter))
+@pytest.fixture(params=data.keys())
+def random_available_min_username(request):
+    return AvailabilityTestCase(
+        request.param,
+        generate_random_username(
+            request.param,
+            websites['username_patterns'][request.param]['available_min_length']))
 
 
-def generate_random_character(length, first_char_letter):
-    for i in range(length):
-        if first_char_letter and i == 0:
-            yield random.choice(string.ascii_letters)
-        else:
-            yield random.choice(string.ascii_letters + string.digits)
+def generate_random_username(website, length):
+    result = ''
+    username_pattern = websites['username_patterns'][website]
+    while True:
+        random_username = exrex.getone('[{chars}]{{{len}}}'.format(
```

Comment:
`for username in exrex.generate(..):` would be better.

currently you have a bug if the generator runs out without yielding a valid username.

with for syntax you can use

```py
for username in exrex.generate(..):

else:
    raise SomeException
```

which properly handles the case where the generator ran out of options, by raising an exception here, not letting the exception occur somewhere else.

Reply:
yep, thanks for finding.

Reply:
Since `exrex.generate` yield value in alphabetic order, I reset it.
@jayvdb, I think that `exrex.generate` generates enough username.
if we use alphabet and generate username of length 3,
total number of cases will be 52^3=140608. 
If we generate length 4,
total number of cases will be 52^47=7311616.
it's pretty big.

Reply:
'min' available usernames are better checked in a systematic manner, not randomly.
But that is not necessary for this task, so if you are finding it difficulty, dont worry about using their generator.  Instead you can use `for` with your own generator which fetches using `getone`.

Reply:
hmm... I can't understand why should I make a generator that fetches by `getone`. what''s the difference between while loop and for loop?
Since `getone` returns(or yields) value randomly, both two ways will fall into infinite loop when there's no valid username exist.

Reply:
Yes, "a generator that fetches by getone" still has the infinite loop bug, which should be noted as a comment, but then that is a bug localised within the generator, and a temporary fix would be for the generator function to use a counter to die before the infinite loop bug occurs.

Using `for: else:` syntax mentioned above means it is designed & implemented correctly, and only the generator needs to be improved to. (and `exrex` have a TODO item to have `generate` be random.)

Reply:
got it.